lib/flux-core/src/iter/traits/iterator.rs

```diff
@@ -0,0 +1,33 @@
+use flux_attrs::*;
+
+#[extern_spec(core::iter)]
+#[assoc(fn valid_item(self: Self, item: Self::Item) -> bool { true })]
+#[assoc(fn size(self: Self) -> int)]
+#[assoc(fn done(self: Self) -> bool)]
+#[assoc(fn step(self: Self, other: Self) -> bool { true })]
+trait Iterator {
+    #[flux::sig(fn(self: &strg Self[@curr_s]) -> Option<Self::Item>[!<Self as Iterator>::done(curr_s)] ensures self: Self{next_s: <Self as Iterator>::step(curr_s, next_s)})]
+    fn next(&mut self) -> Option<Self::Item>;
+
+    #[flux::sig(fn(Self[@s]) -> Enumerate<Self>[0, s])]
+    fn enumerate(self) -> Enumerate<Self>
+    where
+        Self: Sized;
+
+    #[flux::sig(fn(Self[@s], f: F) -> Map<Self, F>[s] where F: FnMut(Self::Item{item: <Self as Iterator>::valid_item(s, item)}) -> B)]
+    fn map<B, F>(self, f: F) -> Map<Self, F>
+    where
+        Self: Sized,
+        F: FnMut(Self::Item) -> B;
+
+    #[flux::sig(fn(Self[@s], f: F) where F: FnMut(Self::Item{item: <Self as Iterator>::valid_item(s, item)}) -> () )]
```

Comment:
We should be consistent with the use of spec

```suggestion
    #[spec(fn(Self[@s], f: F) where F: FnMut(Self::Item{item: <Self as Iterator>::valid_item(s, item)}) -> () )]
```

This should work, but it'll be better to also add `spec` to the list in `flux_attrs_impl::FLUX_ATTRS`.